test/extended/networking/util.go

```diff
@@ -997,10 +995,176 @@ func hasNetworkConfigWriteAccess(oc *exutil.CLI) (bool, error) {
+	allocate := func(template string, n, min, max int, existing sets.Set[string]) ([]string, error) {
+		ips := sets.New[string]()
+		for {
+			if ips.Len() == n {
+				// all requested IPs allocated
+				break
+			}
+			// for each IP to allocate, we take a random start and consecutively
+			// try from there
+			from := rand.Intn(max - min + 1)
+			to := from + max - min
+			var i int
+			for i = from; i < to; i = i + 1 {
+				ip := fmt.Sprintf(template, min+(i%(max-min)))
+				if !ips.Has(ip) && !existing.Has(ip) {
+					ips.Insert(ip)
+					break
+				}
+			}
+			if i == to {
+				return nil, allocationFull
+			}
+		}
+		return ips.UnsortedList(), nil
+	}
+
+	var ipsv4, ipsv6 []string
+	err = retry.OnError(
+		wait.Backoff{Steps: 30, Duration: time.Second},
+		func(err error) bool {
+			return apierrors.IsConflict(err) || errors.Is(err, allocationFull)
+		},
+		func() error {
+			cm, err := oc.AdminKubeClient().CoreV1().ConfigMaps(allocationConfigMapNamespace).Get(context.Background(), allocationConfigMapName, metav1.GetOptions{})
+			if err != nil {
+				return fmt.Errorf("failed to get config map %q: %w", allocationConfigMapName, err)
+			}
+
+			existing := []string{}
+			if cm.Data == nil {
+				cm.Data = map[string]string{}
+			}
+			if cm.Data[key] != "" {
+				err := json.Unmarshal([]byte(cm.Data[key]), &existing)
+				if err != nil {
+					return err
+				}
+			}
+			existingSet := sets.New(existing...)
+			ipsv4, err = allocate(t4, n4, min, max, existingSet)
+			if err != nil {
+				return fmt.Errorf("could not allocate %d IPv4 from pool %s with existin allocations %v: %w", n4, key, existing, err)
+			}
+			ipsv6, err = allocate(t6, n6, min, max, existingSet)
+			if err != nil {
+				return fmt.Errorf("could not allocate %d IPv6 from pool %s with existin allocations %v: %w", n6, key, existing, err)
+			}
+			if len(ipsv4)+len(IPv6) == 0 {
+				return nil
+			}
+			existingSet.Insert(ipsv4...)
+			existingSet.Insert(ipsv6...)
+
+			data, err := json.Marshal(existingSet.UnsortedList())
+			if err != nil {
+				return err
+			}
+
+			cm.Data[key] = string(data)
+			_, err = oc.AdminKubeClient().CoreV1().ConfigMaps(allocationConfigMapNamespace).Update(context.Background(), cm, metav1.UpdateOptions{})
+			if err != nil {
+				return fmt.Errorf("failed to update config map %v: %w", cm, err)
+			}
+			return nil
+		},
+	)
+
+	return ipsv4, ipsv6, err
+}
+
+// deallocateIPs frees IPs previously allocated with allocateIPs
+func deallocateIPs(oc *exutil.CLI, namespace, name, key string, ips ...string) error {
+	return retry.RetryOnConflict(
+		wait.Backoff{Steps: 30, Duration: time.Second},
+		func() error {
+			cm, err := oc.AdminKubeClient().CoreV1().ConfigMaps(allocationConfigMapNamespace).Get(context.Background(), allocationConfigMapName, metav1.GetOptions{})
+			if err != nil {
+				return fmt.Errorf("failed to get config map %q: %w", allocationConfigMapName, err)
+			}
+			existing := []string{}
+			if cm.Data[key] != "" {
```

Comment:
If the `key` doesn't exist, we don't need to update the CM. We can return from here.